Fix: Video description collapsible and no longer hijacks page scroll

sabs21 opened the pull request:
# Fix: Video description collapsible and no longer hijacks page scroll

<!-- Thanks for sending a pull request! Make sure to follow the contributing guidelines. -->
<!-- Important note, we may remove your pull request if you do not use this provided PR template correctly. -->

## Pull Request Type
<!-- Please select what type of pull request this is: [x] -->
- [x] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

## Related issue
<!-- Please link the issue your pull request is referring to. -->
<!-- If this pull request fully resolves the relevant issue, put "closes" before the issue number. -->
<!-- Example: "closes #123456". -->
closes #4821 

## Description
<!-- Please write a clear and concise description of what the pull request does. -->
Long video descriptions will now have their overflow hidden to prevent hijacking the scroll of the page. To see the full description and reintroduce the scrollbar, the user must click 'Click to View Description' shown at the bottom of the description. The 'Click to View Description' button is only shown for descriptions with a computed CSS height greater than or equal to 300 pixels.

## Screenshots <!-- If appropriate -->
<!-- Please add before and after screenshots if there is a visible change. -->
CURRENT:
![image](https://github.com/user-attachments/assets/6df0026b-ec17-4b92-ae6a-5646af9a2c00)

CHANGE (Description not expanded):
![image](https://github.com/user-attachments/assets/c8ee6fa5-233f-470c-8f68-a35b55ef2509)

CHANGE (Description expanded):
![image](https://github.com/user-attachments/assets/2b8d9890-6d4d-44b8-a9df-20f38dbdc58c)


## Testing <!-- for code that is not small enough to be easily understandable -->
<!-- Has this pull request been tested? -->
<!-- Please describe shortly how you tested it. -->
<!-- Are there any ramifications remaining? -->
I tested this change on a video with a long description (description exceeds CSS height of 300px) and a video with a short description. I verified: 
- Clicking 'Click to View Description' makes the description scrollable and that the button removes itself.
- Short descriptions do not show the 'Click to View Description' button.

## Desktop
<!-- Please complete the following information-->
- **OS:** Windows 10
- **OS Version:** 22H2, Build 19045.4842
- **FreeTube version:** v0.21.3 Beta

## Additional context
<!-- Add any other context about the pull request here. -->

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Could you explain why you chose for 300px? Is the behavior the same as YT? If not why is this better?

sabs21 commented:
> Could you explain why you chose for 300px? Is the behavior the same as YT? If not why is this better?

I chose 300px because that's the height/max-block-size the description was already set to (look at .videoDescription in WatchVideoDescription.css).

kommunarr commented:
My (purely visual, untested) feedback: I like YT's UI for this better, particularly that the click area is the full description box so it's easier to click, the "more" button is casual, placed inside the flow of the description, and not too strongly competing for visual attention with an underline (and at least for En-US users, in lower case), and the corresponding "less". Also like that it does so by fully unraveling it rather than just making a scrollable div.

sabs21 commented:
> My (purely visual, untested) feedback: I like YT's UI for this better, particularly that the click area is the full description box so it's easier to click, the "more" button is casual, placed inside the flow of the description, and not too strongly competing for visual attention with an underline (and at least for En-US users, in lower case), and the corresponding "less". Also like that it does so by fully unraveling it rather than just making the div scrollable.

Agreed. I went ahead and implemented your feedback in another branch. Here are screenshots of what I did. If this looks like its on a better track, I'll merge my new branch into this one. 

The only piece here that I didn't do is make the description fully unravel. The reason is that for very long descriptions, the description becomes longer than the recommended videos component on the right. This causes the page to auto load recommended videos which makes the page longer, causing the description to partially scroll back up and makes scrolling to the bottom of the description difficult.

On load:
<img width="958" alt="image" src="https://github.com/user-attachments/assets/0f5c93d9-fa1a-4a12-9761-620dee671c0c">

Hovering over the description (entire description is clickable while collapsed):
<img width="957" alt="image" src="https://github.com/user-attachments/assets/30161ba6-1e9e-4128-ad87-e2050e0b6fbc">

Clicked description:
<img width="956" alt="image" src="https://github.com/user-attachments/assets/370ab45e-5645-4f68-b878-3f155719923a">

Show less:
<img width="956" alt="image" src="https://github.com/user-attachments/assets/0d95944d-6670-49a6-85a4-7a5c03ce8ce2">

efb4f5ff-1298-471a-8973-3d47447115dc commented:
@kommunarr any additional thought on the way its implemented now?

kommunarr commented:
Sorry for the delay, that looks great @sabs21! Please merge that in.

kommunarr reviewed (changes requested):
This is awesome!

**suggestion (blocking):** The main issue I can see is that the "...more" button doesn't work as a button when accessed through tabbing, and the "full box button" does appear in the tab order. My suggestion is to have the "...more" button work when tabbing, and have the "full box button" not appear in the tab order.

kommunarr commented:
Hi @sabs21, is there any way I can help or provide clarification on the issue & recommended fix?

sabs21 commented:
> Hi @sabs21, is there any way I can help or provide clarification on the issue & recommended fix?

Apologies for the delay, admittedly since review has began on this I started a software dev job and programming more afterwards has been the last thing I wanted to do lol. I'm sure the accessibility piece is simply down to adding a "tabindex=0" to the description. I'd have to test it, but I'd assume hitting enter on focus should trigger its click event.

kommunarr commented:
No worries, I've been there before! To clarify, my suggestion is to remove the `tabindex=0` on the full box button (to remove it from the tab order) and move the `keydown` event listeners from the full box button to specifically the "Show more" button.
https://github.com/FreeTubeApp/FreeTube/blob/0a6c6e5c1534e91bb99e3154f1c82f5f8920a23e/src/renderer/components/WatchVideoDescription/WatchVideoDescription.vue#L25-L30

This way, when a user navigates using keyboard navigation, the full box button will not appear in the tab order button, and the "Show more" button will work when tabbing to it and pressing <kbd>Enter</kbd>.

sabs21 commented:
> No worries, I've been there before! To clarify, my suggestion is to remove the `tabindex=0` on the full box button (to remove it from the tab order) and move the `keydown` event listeners from the full box button to specifically the "Show more" button.
> 
> https://github.com/FreeTubeApp/FreeTube/blob/0a6c6e5c1534e91bb99e3154f1c82f5f8920a23e/src/renderer/components/WatchVideoDescription/WatchVideoDescription.vue#L25-L30
> 
> This way, when a user navigates using keyboard navigation, the full box button will not appear in the tab order, and the "Show more" button will work when tabbing to it and pressing Enter.

I'll give this a shot now and see where it goes. Make sense to me on the surface!

sabs21 commented:
Unfortunately I'm battling the eslinter here. I'd like to have the description overlay expand the description on click similar to how YT's behavior is. The problem being that eslint wants me to make the overlay focusable which of course goes against what we want.

I could just push a version without the overlay, but that greatly reduces the clickable area and would likely make the 'more...' button hard to find since it exists on the right side of the description.

kommunarr commented:
Gotcha, if you want to follow YT's implementation for advice, here's how they do it: make the "box button" not an actual button (i.e., remove the role=button), make the "Show more" one have role=button and have no click/keydown handling (which instead handled by bubbling up to its box button parent).

kommunarr commented on src/renderer/components/WatchVideoDescription/WatchVideoDescription.vue:
```diff
@@ -1,19 +1,49 @@
       @timestamp-event="onTimestamp"
     />
+    <template v-if="showControls">
+      <span
+        v-if="showFullDescription"
+        class="descriptionStatus"
+        tabindex="0"
+        @click="collapseDescription"
+        @keydown.space.prevent="collapseDescription"
+        @keydown.enter.prevent="collapseDescription"
+      >
+        {{ $t("Description.Collapse Description") }}
+      </span>
+      <span
+        v-else
+        class="descriptionStatus"
+        tabindex="0"
+        @keydown.space.prevent="expandDescription"
+        @keydown.enter.prevent="expandDescription"
+      >
+        {{ $t("Description.Expand Description") }}
+      </span>
```
This is working great through tab navigation! Last comment; could you add role="button" to these two?

sabs21 replied:
My bad! I accidentally removed those, just pushed that now.

kommunarr reviewed (commented):
LGTM; will approve after the release.

PikachuEXE commented:
I need example video for testing `Short descriptions do not show the 'Click to View Description' button`

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
Pre-approval

kommunarr reviewed (commented):
Discovered one more issue; the "Show more" button is at the bottom of the tab order, so if a video description has links, you will have to tab through all of them to actually click the "Show more" button. It should instead be at the top of the tab order for the container. One possible solution is to [use the CSS order property](https://developer.mozilla.org/en-US/docs/Web/CSS/order). [Here's a video with links in the description as an example.](https://youtu.be/N43mwATceXc)

sabs21 commented:
> Discovered one more issue; the "Show more" button is at the bottom of the tab order, so if a video description has links, you will have to tab through all of them to actually click the "Show more" button. It should instead be at the top of the tab order for the container. One possible solution is to [use the CSS order property](https://developer.mozilla.org/en-US/docs/Web/CSS/order). [Here's a video with links in the description as an example.](https://youtu.be/N43mwATceXc)

Great idea, I wound up flipping the ordering such that the buttons came first in the HTML, then used flexbox + order to adjust physical placement while keeping the desired tab ordering.

kommunarr commented:
Pre-existing: seeing template errors for videos with no description ([ex](https://youtu.be/PNo3yQ-h_ZQ)) due to no use of the null access operator.

For reference, [here's a video](https://youtu.be/floHBfe5-YM) with a short description, where it properly does not show the "Show more" button.

**issue:** Sorry for all the different rounds of feedback, but I think this should be the last! I don't mind the "Show less" button being at the top, but it should have the same top padding as the `p.description` that it replaces.

sabs21 commented:
I went ahead and changed the visual ordering of the show less button so that the padding is normal once again. I didn't notice that the show less button was ordered differently due to me messing around with CSS locally in the dev tools. The spacing now appears normal once again too.

![image](https://github.com/user-attachments/assets/1f2018ee-b9e4-4053-b9b9-1814f16ff889)

kommunarr commented:
As a byproduct of this change, clicking the Show more button now opens at the end of the container (i.e., scrolled all the way down) instead of at the top.

sabs21 commented:
> As a byproduct of this change, clicking the Show more button now opens at the end of the container (i.e., scrolled all the way down) instead of at the top.

I reverted the previous change and just added the margin to the top of the show less button

PikachuEXE commented:
Anyone got a video with short description for testing?

PikachuEXE commented:
Just got error when trying to find a video with short description
https://youtu.be/ASeJV9KGAOc

![image](https://github.com/user-attachments/assets/c3a0a9d5-2e6a-43f5-a79b-0d9522bf2201)

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Show Less is now seen in the top of the description box but i expected it to be shown. Was this an intentional change?

sabs21 commented:
> Show Less is now seen in the top of the description box but i expected it to be shown. Was this an intentional change?

No I didn't expect that to happen, I'll work on fixing that now

sabs21 commented:
Just fixed that, my guess is that when I merged the latest development, there was some css that caused the show less button to appear at the top.

efb4f5ff-1298-471a-8973-3d47447115dc commented:
I cant remember `...more` to be this far to the right. Did this also break?

![image](https://github.com/user-attachments/assets/8326e069-9df8-47ba-89ca-522b6159cdd3)

sabs21 commented:
> I cant remember `...more` to be this far to the right. Did this also break?
> 
> ![image](https://private-user-images.githubusercontent.com/73130443/393498609-8326e069-9df8-47ba-89ca-522b6159cdd3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2MzAxMDgsIm5iZiI6MTczMzYyOTgwOCwicGF0aCI6Ii83MzEzMDQ0My8zOTM0OTg2MDktODMyNmUwNjktOWRmOC00N2JhLTg5Y2EtNTIyYjYxNTljZGQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA4VDAzNTAwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUzNmNjZTU2Y2RjOWJmYTg0OGNhNWIwMmUxNzEzMGVhNTc5MDY0MzFhNWJiNTM1MTg2YjYxN2Q5MWZjMjk5ZWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DaLI11iJoFWx1V1Vx3H7tzlBUuOiM8FmdpwMUc3-L-8)

No, that hasn't changed. You can see this same padding in a screenshot within my September 16th post https://github.com/FreeTubeApp/FreeTube/pull/5665#issuecomment-2354046921

efb4f5ff-1298-471a-8973-3d47447115dc commented:
When expanding the description it scrolls all the way to the bottom instead of starting at top


https://github.com/user-attachments/assets/f83e2faf-dcba-4e03-ab68-cc01c5884046



Noticed a difference on FT vs YT

YT you can highlight text and open the link in the preview

https://github.com/user-attachments/assets/2dd6f0ea-4374-4ceb-8e43-7d7901417271

This isnt possible within FT. I can only highlight all the text within the collapsed container


https://github.com/user-attachments/assets/9755f161-9057-48f4-8a02-27abaf0be34b

Not able to highlight text is minor but maybe the url is?

PikachuEXE commented:
@efb4f5ff-1298-471a-8973-3d47447115dc https://github.com/FreeTubeApp/FreeTube/pull/6461
